src/main/java/com/jwplayer/southpaw/util/S3Helper.java

```diff
@@ -163,33 +215,49 @@ public static String getPath(File file) {
-        Preconditions.checkArgument(SCHEME.equalsIgnoreCase(s3Uri.getScheme()));
-        List<S3ObjectSummary> summaries = listKeys(s3Uri);
-        String bucket = s3Uri.getHost();
-        TransferManager tx = TransferManagerBuilder.standard().withS3Client(s3).build();
-        Map<File, Download> downloads = new HashMap<>();
-        logger.info(String.format("Downloading files from %s to %s", s3Uri.toString(), localUri.toString()));
-        int downloadCount = 0;
+    public void syncFromS3(URI localUri, URI s3Uri) throws InterruptedException, ExecutionException {
+        waitForSyncToS3();
+        try(Timer.Context context = metrics.s3Downloads.time()) {
+            Preconditions.checkNotNull(localUri);
+            Preconditions.checkNotNull(s3Uri);
+            Preconditions.checkArgument(localUri.getScheme() == null || FileHelper.SCHEME.equalsIgnoreCase(localUri.getScheme()));
+            Preconditions.checkArgument(SCHEME.equalsIgnoreCase(s3Uri.getScheme()));
+            List<S3ObjectSummary> summaries = listKeys(s3Uri);
+            String bucket = s3Uri.getHost();
+            TransferManager tx = TransferManagerBuilder.standard().withS3Client(s3).build();
+            Map<File, Download> downloads = new HashMap<>();
+            logger.info(String.format("Downloading files from %s to %s", s3Uri.toString(), localUri.toString()));
+            int downloadCount = 0;
 
-        // Download the files from S3
-        for(S3ObjectSummary summary: summaries) {
-            String suffix = StringUtils.substringAfter(summary.getKey(), getPath(s3Uri));
-            File file = new File("/" + getPath(localUri) + suffix);
+            // Download the files from S3
+            for (S3ObjectSummary summary : summaries) {
+                String suffix = StringUtils.substringAfter(summary.getKey(), getPath(s3Uri));
+                File file = new File("/" + getPath(localUri) + suffix);
 
-            file.getParentFile().mkdirs();
-            downloads.put(file, tx.download(bucket, summary.getKey(), file));
-            downloadCount++;
-            if(downloadCount % MAX_KEYS_PER_S3_OP == 0) {
-                for(Map.Entry<File, Download> download: downloads.entrySet()) {
+                file.getParentFile().mkdirs();
+                downloads.put(file, tx.download(bucket, summary.getKey(), file));
+                downloadCount++;
+                metrics.s3FilesDownloaded.mark(1);
+                if (downloadCount % MAX_KEYS_PER_S3_OP == 0) {
+                    for (Map.Entry<File, Download> download : downloads.entrySet()) {
+                        download.getValue().waitForCompletion();
+                        download.getKey().setLastModified(
+                                download.getValue().getObjectMetadata().getLastModified().getTime()
+                        );
+                    }
+                    logger.info(String.format("Downloaded %s files", downloadCount));
```

Comment:
Minor, but this same log message appears on line 266. Would it be clearer to only print it once at line 268 after all downloads are done

Reply:
The thinking originally is that this leaves a trace of the progress of the uploads/downloads in the logs. Though since the metrics exist now, those provide a good measure of the progress/rates of these S3 syncs.

Reply:
oh missed this being called multiple times within the loop. I thought it was just a full count of downloaded files meant to print once.